src/book/02-system/01-foundations/07-director.mdx

```diff
@@ -0,0 +1,231 @@
+---
+section: System
+chapter: Foundations
+title: Director
+description: Introduction to the Director and how to use it to build behaviours.
+slug: /system/foundations/director
+---
+
+The NUbots use the director system to control all their behaviour elements.
+It allows composable behaviours that can take advantage of transition rules to ensure that the robot performs the right task at the right time.
+
+The Director does this by providing a set of new DSL keywords to NUClear that describe what tasks to run, what conditions are required, and how those tasks build on other tasks.
+It then uses these keywords to make a graph that it will use to decide which tasks to execute in order to achive the goals of the modules.
+
+The two DSL components that are added are Providers and Tasks.
+
+## Providers
+
+Providers are implemented as NUClear reactions and live in a NUClear `on` statement.
+They are the components which "provide" the functionality needed to perform a task, either by doing it directly or by running a series of subtasks.
+They form the basis of the modules that implement behaviours.
+Each provider will provide for a specific message type in a specific situation.
+For example a provider which executed a walk task, the code might look like:
+
+```cpp
+on<Provides<Walk>>().then(...);
+```
+
+Provider are triggered whenever a task needs to be executed by the director, however they can also be triggered using other NUClear keywords.
+The director will ensure that any provider which is not currently active will not run if it is not active, even if the other triggers would make it run.
+For example if the walk engine module wants to update at 10Hz when it is running a provider may look like.
+
+```
+on<Provides<Walk>, Every<10, Per<std::chrono::seconds>>().then(...);
+```
+
+This provider would then run once every time the walk command was updated, as well as running at 10Hz while the provider is active.
+
+If you do want a provider to run even when it is inactive you can make them optional (e.g. `on<Optional<Provides<Walk>>>().then(...);`).
+Then this will still run although its task will be empty.
+Non active providers are unable to emit subtasks and will show an error if they attempt to.
+
+Note that for the director to recognise subtasks as coming from a specific provider, those tasks MUST be emitted from that specific reaction.
+Otherwise they will be considered root tasks and can cause all sorts of chaos.
+
+### Provider Groups
+
+When there are multiple providers which provide for a single task type they are considered a provider group.
```

Comment:
So multiple providers for a single task has heaps of use cases, those you suggested can surely be one if you want it to be.
The more general case I was thinking of is when you have things like a normal walk engine provider
`on<Provides<Walk>>`
And another provider that makes the robot stand by stopping walking
`on<Leaving<Walk>, Causing<Stability::STAND>`

Both of those providers take a walk command and get the walk engine to do something.
When you want to run either of those depends on what you're doing.
If a higher priority task wants to run then it can take advantage of these other providers to get the robot into a state that it wants to (e.g. stand before kick) without you having to program in specialised state control variables to ensure the robot has stopped walking before the kick happens

You can take this concept super far though by making multiple providers that run in different circumstances depending on how crazy you wan to go

Reply:
Right but still "Do all in a group have to have some sort of priority in case of an unknown order?"
How does it decide which one if it isn't obvious? and they're siblings?

Reply:
```suggestion
When there are multiple Providers which provide for a single task type they are considered a provider group.
```